### PR TITLE
feat(research/training): tokenizer scaffold with manifest spine + smoke-validated DDP loop

### PR DESCRIPTION
**Pull request by koriyoshi2041**
## Summary

First real code under `research/training/`. Lands the Phase-1.1 tokenizer training subprogram per `docs/research/2026-05-13-wittgenstein-research-program.md` §1.1 plus the shared infrastructure both adapter (1.2) and llm-head (1.3) will reuse.

Pairs with #491 (Gates C+D audit — provides the LlamaGen baseline to compare own-trained checkpoints against) and #492 (LlamaGen bridge manifest — the audited fallback while own-trained weights train).

### Shared infra (`research/training/_shared/`)

- **`manifest.py`** — `TrainingManifest` emitter, sibling of the inference-side `RunManifest`. Captures runtime fingerprint (git SHA, lockfile SHA, torch/CUDA/cudnn versions, A800 device name), dataset fingerprint (root SHA over sorted file enumeration), optimizer state hash, checkpoint metadata + weights SHA-256, eval snapshot slot, and a full config dump. Atomic write. Matches the receipt spine #441 ratified.
- **`dataset.py`** — `ImageFolderDataset` (index-mappable, robust to corrupt files), `SyntheticImageDataset` (deterministic, for smoke), and the LlamaGen-canonical preprocessing helper so trained-tokenizer outputs sit in the same statistical regime as the audited baseline.
- **`eval.py`** — Rung-1 reconstruction eval harness (PSNR / SSIM / LPIPS / rFID via clean-fid / codebook-usage histogram). Lazy imports so partial dep coverage doesn't kill the eval.
- **`_third_party/llamagen_vq_model.py`** — vendored MIT-licensed architecture from `FoundationVision/LlamaGen @ ce98ec41` (the audited commit). `LICENSE.llamagen` + README document attribution and the re-vendor protocol.

### Tokenizer subprogram (`research/training/tokenizer/`)

- **`model.py`** — factory wrapping the vendored `VQModel` with config-driven **D=32** (vs LlamaGen's D=8) for richer per-site latents under our learned MaskGIT adapter.
- **`losses.py`** — `TokenizerLoss` assembling L2 + LPIPS + GAN-when-on + commit pass-through. Discriminator slot pre-wired.
- **`config.py`** — `TrainConfig` + `smoke_config()`; receipt-friendly snapshot.
- **`train.py`** — **plain-PyTorch DDP-native** training loop per #441 recommendation. No Lightning lock-in. `torchrun` launch. Broadcasts `run_id` from rank 0 so all ranks agree on the manifest path. Emits a Wittgenstein `TrainingManifest` at every checkpoint.
- **`smoke_test.py`** — 5-step end-to-end validation.

## Receipts (cluster-measured, proving this is real, not just code)

| Run | Hardware | Outcome |
|---|---|---|
| Single-GPU smoke | 1× A800-80GB | 71.8M-param model built, 5 steps in 5.2s, manifest written, L2 loss decreased monotone |
| 2× A800 DDP smoke (NCCL) | 2× A800-80GB | Same, scaled to world=2; rank-id broadcast verified |
| **500-step real-data validation** | 2× A800-80GB, NCCL | L2 **0.4935 → 0.0473** (lowest at step 400), codebook usage **8.5% → 25%**, **72 imgs/s** sustained. Final checkpoint **SHA matches manifest exactly** (byte-pinned). Run: `tokenizer-20260527T101458Z-dac1ce6e` |

The 500-step run used NExTQA video frames (20,155 images, research-only license — scaffold validation only, NOT a publishable checkpoint).

## What this PR does NOT do (intentionally deferred)

- **Does NOT launch the publishable Phase-1.1 training run.** Full ImageNet + CC12M training needs (a) DVC remote (#400) + dataset on cluster, (b) training-stack re-audit close (#441), (c) maintainer sign-off via the model-owner hub (#435).
- **Does NOT implement the PatchGAN discriminator.** `losses.py` is shaped for it; first PR keeps the GAN off so the recon-only trajectory is observable on its own. Adding GAN is a small follow-up.
- **Does NOT add resume-from-checkpoint, aim/W&B integration, or DVC dataset hashing.** The receipt slots exist (`manifest.experiment_uri`, `dataset.revision`); follow-ups fill them.
- **Does NOT touch the adapter (Phase 1.2) or llm-head (Phase 1.3) beyond README polish** — those land after Phase 1.1 ships a real checkpoint.

## Test plan

- [ ] `scripts/check-no-research-imports.mjs` green — code only imports `_shared/_third_party/llamagen_vq_model` from inside `research/training/`; never reaches into `packages/`
- [ ] `scripts/check-npm-publish-tarball.mjs` green — research/ is outside the publish surface (this PR adds nothing under packages/)
- [ ] `pnpm typecheck` / `pnpm test` unchanged (additive only, no TS imports affected)
- [ ] (manual, contributor-local) `python -m research.training.tokenizer.smoke_test` exits 0 and writes a `final.manifest.json`
- [ ] (manual, GPU contributor) `torchrun --nproc-per-node N -m research.training.tokenizer.smoke_test` exits 0 across N ranks
- [ ] Two-hat review per `docs/tracks.md`: researcher (does D=32 + receipt shape match research-program §1.1?) + hacker (would a 2 a.m. agent extend `_shared/manifest.py` without leaking out of receipt discipline?)

Refs #70, #283, #396, #397, #441

### CODE REVIEW

**Comment by Jah-yee** [comment]
Maintainer gate note before merge:

This PR is training-adjacent in a strong sense: it adds the tokenizer scaffold, DDP-native smoke loop, training manifests, vendored LlamaGen VQ model code, and cluster-measured smoke/500-step validation claims. I added `needs-ml-specialist` so the queue reflects that this should not be treated as an engineering-only scaffold.

Two review boundaries I think we should preserve:

- Engineering review can approve package isolation, manifest shape, deterministic smoke mechanics, CI failures, vendoring hygiene, and publish-surface safety.
- Research / ML review should independently sign off on the D=32 tokenizer choice, eval metric shape, dataset/license framing, DDP/manifest claims, and whether the reported smoke receipts are sufficient for this PR's stated scope.

Also note that the current CodeRabbit comment is rate-limit/credit exhaustion, not a substantive completed review, and Static checks are red on `typos` for the vendored `nin_shortcut` symbol. Suggested small fix: allowlist that upstream symbol or exclude the vendored file from typos rather than renaming vendored code.

**Comment by Jah-yee** [comment]
Engineering review note before this gets treated as a generic "training is ready" signal:

I think this PR is worth reviewing, but the attention split should stay sharp.

**Engineering surfaces that are reviewable now**
- vendoring hygiene for `research/training/_shared/_third_party/`
- manifest shape / receipt-writing mechanics
- smoke-loop determinism and DDP wiring
- publish-surface isolation and repo-boundary discipline
- CI hygiene, including the current `typos` failure

**Surfaces that still need independent ML / research judgment**
- D=32 tokenizer choice versus the audited LlamaGen D=8 baseline
- dataset/license framing for the reported 500-step validation run
- whether the eval stack and reported smoke receipts are sufficient for the claims in this PR body
- whether this scaffold is the right Phase 1.1 starting point versus just a useful prototype

One concrete engineering ask before merge: the current red Static check should be fixed by allowlisting or excluding the vendored upstream symbol `nin_shortcut` rather than editing the vendored file. That keeps provenance clean and avoids inventing a Wittgenstein-only fork of upstream code for a spelling tool.

@moapacha if you have bandwidth, a short maintainer comment on package isolation / vendoring / receipt discipline would help separate "engineering scaffold passes" from "training approach endorsed" for later readers.

**Comment by Jah-yee** [comment]
Maintainer note: from the engineering side, the main deterministic fix I would still apply here is to resolve the current static-check noise without touching vendored upstream model code. The vendored `llamagen_vq_model.py` contains the upstream symbol `nin_shortcut`; that looks like a false positive from typo scanning, not a code defect.

I prepared a minimal local fix (`387f979`) that allowlists `nin_shortcut` in `typos.toml` instead of renaming the vendored symbol. That keeps the vendored copy faithful to upstream and avoids creating meaningless diff churn in a training-adjacent third-party file.

Reviewer focus suggestion: engineering review here should stay narrow—package isolation, manifest/receipt shape, smoke mechanics, vendoring hygiene, publish-surface boundaries. ML/research sign-off is still required for tokenizer choice, architecture claims, eval interpretation, dataset/license posture, and any “validated training” conclusions.

I would treat the allowlist patch as safe to land independently of the broader research review.

**Comment by Jah-yee** [comment]
Maintainer note: I pushed a narrow follow-up PR that captures the remaining engineering closeout I think is appropriate for this line: #517.

Scope is intentionally limited to engineering acceptance and observability, not model-quality blessing. It does two things: (1) allowlists the vendored upstream symbol `nin_shortcut` instead of mutating third-party model code, and (2) tightens the tokenizer scaffold’s minimum acceptance surface so smoke / receipt / degraded-eval states are easier to review explicitly.

Concretely, the follow-up makes smoke verify acceptance facts beyond “a manifest file exists”, records degraded eval coverage in receipt metrics instead of only printing stdout warnings, and documents a minimum pass list for reviewers in the tokenizer README.

I think this is the right boundary for engineering review here: loop integrity, manifest integrity, fallback visibility, and publish-surface hygiene. It still does **not** turn smoke success into tokenizer-quality approval, and it does **not** replace independent ML/research sign-off for architecture choice, metric thresholds, dataset posture, or “validated training” claims.

**Comment by moapacha** [comment]
**CI triage: the Static-checks failure is stale — just rebase.**

The red `Static checks (spelling + workflow + deps)` run on this PR is from 2026-05-27T16:21, which **predates** [#514](https://github.com/p-to-q/wittgenstein/pull/514) (merged 2026-05-28). #514 added `nin = "nin"` to `typos.toml` repo-wide — exactly the `nin_shortcut` sub-token this PR's vendored tokenizer code trips on.

Verified: this branch does **not** re-add `nin` itself, so once you rebase onto current `main` the spelling gate passes with no change to your PR.

```
git fetch origin
git rebase origin/main
git push --force-with-lease
```

That also picks up [#518](https://github.com/p-to-q/wittgenstein/pull/518) (publish-tarball guard now fences `python/` + `polyglot-mini/` and fixed the dead `benchmarks/` pattern) — relevant since this PR adds training Python; the guard will confirm none of it leaks into the cli tarball.

No content change needed from you for the spelling gate. The substantive tokenizer-scaffold review stays with the ML-owner track (you're the owner here) — this note is just unblocking the mechanical CI red.

**Comment by Jah-yee** [comment]
Maintainer note: after `#517` merged, the main remaining blocker on this parent PR is now very narrow and mechanical: `Static checks` is red because the `typos.toml` merge context ends up with a duplicate `nin = "nin"` entry. This is not a new training/eval concern; it is a follow-on cleanup from the allowlist work moving into branch ancestry.

I opened a minimal follow-up PR for exactly that fix: #522. Once that lands, this parent PR should be back to a much cleaner state from the engineering side.

**Comment by Jah-yee** [comment]
Maintainer follow-up: I tested the alternate hypothesis via #522. The result is clear: dropping `nin = "nin"` is **not** a valid fix, because `typos` immediately re-flags the vendored `nin_shortcut` sub-token. So the right end state for this line is: keep **one** `nin = "nin"` entry, and remove any duplicate copy that appears in the merge context. `nin_shortcut` is not the right canonical allowlist target here.

**Comment by Jah-yee** [comment]
Maintainer follow-up: I rebased this branch onto current `main`, kept the final `typos` allowlist in the minimal documented form, and pushed `98ee57b`.

Engineering closeout from my side is now intentionally narrow and complete:
- keep the vendored upstream `nin_shortcut` symbol untouched
- keep exactly one `nin = "nin"` allowlist entry
- document why that allowlist exists so later cleanup does not regress CI
- preserve the broader boundary that smoke / receipt integrity is reviewable here, while tokenizer-quality conclusions remain a research/ML judgment

At this point the remaining signal to watch is just the fresh CI run on `98ee57b` (`Verify (Node + Python)` / site build). If those finish green, I consider the engineering concerns on this PR closed.
